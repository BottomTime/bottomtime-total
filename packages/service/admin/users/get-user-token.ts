/* eslint-disable no-process-env, no-console */
import { UserEntity } from '@/data';

import { getDataSource } from 'admin/database/data-source';
import jwt from 'jsonwebtoken';
import { DataSource } from 'typeorm';

export async function getUserToken(postgresUri: string, username: string) {
  let ds: DataSource | undefined;

  try {
    ds = await getDataSource(postgresUri);

    const lowered = username.trim().toLowerCase();
    const users = ds.getRepository(UserEntity);

    const user = await users.findOne({
      where: [{ usernameLowered: lowered }, { emailLowered: lowered }],
      select: ['id'],
    });

    if (user) {
      const token = await new Promise<string>((resolve, reject) => {
        jwt.sign(
          {
            iss: process.env.BT_BASE_URL,
            iat: Date.now(),
            exp: Date.now() + 20 * 60 * 1000,
            sub: `user|${user.id}`,
          },
          process.env.BT_SESSION_SECRET ?? '',
          {},
          (error, token) => {
            if (error) reject(error);
            else resolve(token!);
          },
        );
      });
      console.log(token);
    } else {
      console.error(`User not found: "${username}"`);
      process.exit(1);
    }
  } catch (error) {
    console.error('Error getting user token:', error);
    process.exitCode = 1;
  } finally {
    if (ds) await ds.destroy();
  }
}
