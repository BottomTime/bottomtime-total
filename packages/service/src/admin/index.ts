export * from './admin.service';

export * from './admin-tanks.controller';
export * from './admin-users.controller';

export * from './admin.module';
