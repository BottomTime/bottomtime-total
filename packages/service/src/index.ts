import { Config } from './config';
import { createApp } from './create-app';
import { createLogger } from './logger';

const log = createLogger(Config.logLevel);

createApp(log)
  .then(async (app) => {
    await app.listen(Config.port);

    log.info(
      `ðŸŽ‰ Service has successfully started and is listening on port ${Config.port}. ðŸŽ‰`,
    );
  })
  .catch((error) => {
    log.fatal('[SERVICE] Failed to start service... Shutting down...', error);
    process.exit(1);
  });
