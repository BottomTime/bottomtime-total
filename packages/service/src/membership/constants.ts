export const ProFeature = 'pro-features';
export const ShopOwnerFeature = 'shop-owner-features';
