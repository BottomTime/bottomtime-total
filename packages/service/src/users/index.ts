export * from './assert-target-user.guard';
export * from './assert-account-owner.guard';
export * from './notification';
export * from './notifications.service';
export * from './user';
export * from './users.service';
export * from './users.controller';
export * from './users.module';
