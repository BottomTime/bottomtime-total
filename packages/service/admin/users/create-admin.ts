import {
  PasswordStrengthSchema,
  ProfileVisibility,
  UserRole,
  UsernameSchema,
} from '@bottomtime/api';

import { hash } from 'bcrypt';
import prompts from 'prompts';
import { UserEntity } from 'src/data';
import { DataSource } from 'typeorm';
import { v4 as uuid } from 'uuid';
import { z } from 'zod';

import { getDataSource } from '../database/data-source';

/* eslint-disable no-console */
export type CreateAdminOptions = {
  potgresUri: string;
  username?: string;
  password?: string;
  email?: string;
};

async function completeOptions(
  options: CreateAdminOptions,
): Promise<Required<CreateAdminOptions>> {
  let username: string;
  let password: string;
  let email: string;

  if (!options.username) {
    const response = await prompts(
      {
        type: 'text',
        name: 'username',
        message: 'Enter a username for the new admin account:',
        validate: (username) => {
          const valid = UsernameSchema.safeParse(username);
          if (valid.success) return true;
          else
            return 'Username must be between 3 and 50 characters and be alphanumeric with dashes, dots, or underscores.';
        },
      },
      { onCancel: () => process.exit(1) },
    );
    username = response.username;
  } else {
    username = options.username;
  }

  if (!options.password) {
    const response = await prompts(
      {
        type: 'password',
        name: 'password',
        message: 'Enter a password for the new admin account:',
        validate: (password) => {
          const valid = PasswordStrengthSchema.safeParse(password);
          if (valid.success) return true;
          else
            return 'Password must be between 8 and 50 characters and contain at least one uppercase letter, one lowercase letter, one digit, and one special character.';
        },
      },
      { onCancel: () => process.exit(1) },
    );
    password = response.password;
  } else {
    password = options.password;
  }

  if (!options.email) {
    const response = await prompts(
      {
        type: 'text',
        name: 'email',
        message: 'Enter an email address for the new admin account:',
        validate: (email) => {
          const valid = z.string().email().safeParse(email);
          if (valid.success) return true;
          else return 'Invalid email address';
        },
      },
      { onCancel: () => process.exit(1) },
    );
    email = response.email;
  } else {
    email = options.email;
  }

  return {
    username,
    password,
    email,
    potgresUri: options.potgresUri,
  };
}

export async function createAdmin(options: CreateAdminOptions): Promise<void> {
  const completedOptions = await completeOptions(options);
  let dataSource: DataSource | undefined;

  try {
    console.log('Connecting to database...');
    dataSource = await getDataSource(options.potgresUri);

    console.log('Creating admin account...');
    const admin = new UserEntity();
    admin.email = completedOptions.email;
    admin.emailLowered = admin.email.toLowerCase();
    admin.emailVerified = true;
    admin.id = uuid();
    admin.name = 'Administrator';
    admin.passwordHash = await hash(completedOptions.password, 16);
    admin.profileVisibility = ProfileVisibility.Private;
    admin.role = UserRole.Admin;
    admin.username = completedOptions.username;
    admin.usernameLowered = admin.username.toLowerCase();

    const Users = dataSource.getRepository(UserEntity);
    await Users.save(admin);

    console.log('Admin account has been saved.');
  } catch (error) {
    console.error(error);
    process.exit(1);
  } finally {
    if (dataSource) await dataSource.destroy();
  }
}
