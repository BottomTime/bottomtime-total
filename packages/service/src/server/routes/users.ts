import { Express } from 'express';
import Logger from 'bunyan';
import { NextFunction, Request, Response } from 'express';
import { z } from 'zod';

import { UserRole } from '../../constants';
import {
  ForbiddenError,
  MissingResourceError,
  UnauthorizedError,
} from '../../errors';
import { assertValid } from '../../helpers/validation';
import {
  CreateUserOptions,
  CreateUserOptionsSchema,
  PasswordStrengthSchema,
  SearchUsersOptions,
  SearchUsersOptionsSchema,
  User,
} from '../../users';
import {
  MailClient,
  ResetPasswordEmailTemplate,
  WelcomeEmailTemplate,
} from '../../email';
import { VerifyEmailTemplate } from '../../email/verify-email-template';
import { requireAdmin, requireAuth } from './auth';
import { issueAuthCookie } from '../jwt';
import { UsernameSchema } from '../../data';

const VerifyEmailBodySchema = z.object({
  token: z.string().trim(),
});
type VerifyEmailParams = z.infer<typeof VerifyEmailBodySchema>;

const ResetPasswordBodySchema = z.object({
  token: z.string().trim(),
  newPassword: PasswordStrengthSchema,
});
type ResetPasswordParams = z.infer<typeof ResetPasswordBodySchema>;

async function sendWelcomeEmail(
  mail: MailClient,
  user: User,
  log: Logger,
): Promise<void> {
  try {
    if (log.debug()) {
      log.debug(`Requesting email verification token for "${user.email}"...`);
    }
    const verifyEmailToken = await user.requestEmailVerificationToken();

    if (log.debug())
      log.debug(`Generating welcome email body for "${user.email}"...`);
    const mailTemplate = new WelcomeEmailTemplate();
    const messageBody = await mailTemplate.render({
      user,
      verifyEmailToken,
    });

    if (log.debug()) log.debug(`Sending welcome email to "${user.email}"...`);
    await mail.sendMail(
      { to: user.email! },
      'Welcome to Bottom Time!',
      messageBody,
    );
  } catch (mailError) {
    log.error(
      `Failed to send welcome email to new user: ${user.username}`,
      mailError,
    );
  }
}

export async function createUser(
  req: Request,
  res: Response,
  next: NextFunction,
) {
  try {
    const { username } = req.params;
    const options = assertValid<CreateUserOptions>(
      {
        username,
        ...req.body,
      },
      CreateUserOptionsSchema,
    );

    if (options.role && options.role !== UserRole.User) {
      const errorMessage =
        'Unable to create new user account with elevated privileges. Only authenticated admins are able to do that.';
      if (!req.user) {
        throw new UnauthorizedError(errorMessage);
      }
      if (req.user.role !== UserRole.Admin) {
        throw new ForbiddenError(errorMessage);
      }
    }

    const user = await req.userManager.createUser(options);

    if (req.log.info()) {
      req.log.info('New user account created.', {
        username: user.username,
        email: user.email,
      });
    }

    // Sign in the new user if they are not currently logged in as someone else.
    if (!req.user) {
      await issueAuthCookie(user, res);
      req.log.info(
        `User has been logged into their newly-created account: "${user.username}"`,
      );
    }

    // Send the new user a welcome email.
    if (user.email) {
      await sendWelcomeEmail(req.mail, user, req.log);
    }

    res.status(201).json(user);
  } catch (error) {
    next(error);
  }
}

export async function getUserExists(req: Request, res: Response) {
  res.sendStatus(200);
}

export async function getUser(req: Request, res: Response) {
  res.json(req.selectedUser);
}

export async function loadUserAccount(
  req: Request,
  _res: Response,
  next: NextFunction,
) {
  try {
    req.selectedUser = await req.userManager.getUserByUsernameOrEmail(
      req.params.username,
    );
    if (req.selectedUser) {
      next();
    } else {
      next(
        new MissingResourceError(
          `No user found with username: ${req.params.username}`,
        ),
      );
    }
  } catch (error) {
    next(error);
  }
}

export async function requestPasswordResetEmail(
  req: Request,
  res: Response,
  next: NextFunction,
) {
  try {
    const user = await req.userManager.getUserByUsernameOrEmail(
      req.params.username,
    );
    if (!user || !user.email) {
      res.sendStatus(204);
      return;
    }

    req.log.debug(
      `Requesting password reset token for user "${user.username}"...`,
    );
    const mailTemplate = new ResetPasswordEmailTemplate();
    const resetToken = await user.requestPasswordResetToken();

    req.log.debug(`Generating and sending password reset email....`);
    const messageBody = await mailTemplate.render({
      user,
      resetToken,
    });
    await req.mail.sendMail(
      { to: user.email },
      'Reset your Bottom Time password',
      messageBody,
    );

    res.sendStatus(204);
  } catch (error) {
    next(error);
  }
}

export async function requestVerificationEmail(
  req: Request,
  res: Response,
  next: NextFunction,
) {
  try {
    const user = req.selectedUser;
    if (!user?.email) {
      // Fail silently if the user does not have an email address set.
      res.json({ succeeded: false });
      return;
    }

    const mailTemplate = new VerifyEmailTemplate();
    const verifyEmailToken = await user.requestEmailVerificationToken();

    const messageBody = await mailTemplate.render({
      user,
      verifyEmailToken,
    });
    await req.mail.sendMail(
      { to: user.email },
      'Verify your email address',
      messageBody,
    );

    res.sendStatus(204);
  } catch (error) {
    next(error);
  }
}

export async function resetPassword(
  req: Request,
  res: Response,
  next: NextFunction,
) {
  try {
    if (!req.selectedUser) {
      throw new Error('No user profile loaded.');
    }

    const { token, newPassword } = assertValid<ResetPasswordParams>(
      req.body,
      ResetPasswordBodySchema,
    );
    const succeeded = await req.selectedUser.resetPassword(token, newPassword);

    res.json({ succeeded });
  } catch (error) {
    next(error);
  }
}

export async function searchUsers(
  req: Request,
  res: Response,
  next: NextFunction,
) {
  try {
    const options = assertValid<SearchUsersOptions>(
      req.query,
      SearchUsersOptionsSchema,
    );

    const users = await req.userManager.searchUsers(options);
    res.json({
      count: users.length,
      results: users.map((user) => user.toJSON()),
    });
  } catch (error) {
    next(error);
  }
}

export async function verifyEmail(
  req: Request,
  res: Response,
  next: NextFunction,
) {
  try {
    const user = req.selectedUser!;
    const { token } = assertValid<VerifyEmailParams>(
      req.body,
      VerifyEmailBodySchema,
    );

    const verified = await user.verifyEmail(token);

    res.json({ verified });
  } catch (error) {
    next(error);
  }
}

export function configureUserRoutes(app: Express) {
  const UserRoute = '/users/:username';
  app.get('/users', requireAdmin, searchUsers);
  app
    .route(UserRoute)
    .head(loadUserAccount, getUserExists)
    .get(requireAdmin, loadUserAccount, getUser)
    .put(createUser);
  app.post(
    `${UserRoute}/requestEmailVerification`,
    requireAuth,
    loadUserAccount,
    requestVerificationEmail,
  );
  app.post(`${UserRoute}/requestPasswordReset`, requestPasswordResetEmail);
  app.post(`${UserRoute}/resetPassword`, loadUserAccount, resetPassword);
  app.post(
    `${UserRoute}/verifyEmail`,
    requireAuth,
    loadUserAccount,
    verifyEmail,
  );
}
