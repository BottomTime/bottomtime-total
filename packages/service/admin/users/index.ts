import { CommandModule } from 'yargs';

import { createAdmin } from './create-admin';
import { getUserToken } from './get-user-token';

export const userModule: CommandModule<{
  'postgres-uri': string;
  'require-ssl': boolean;
}> = {
  command: 'user',

  describe: 'Commands for managing a user account',

  builder(yargs) {
    return yargs
      .command(
        'new-admin',
        'Creates a new admin user.',
        (yargs) => {
          return yargs
            .positional('username', {
              alias: 'u',
              description: "The user's username.",
              type: 'string',
            })
            .positional('password', {
              alias: 'p',
              description: 'The password to set on the new admin account',
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await createAdmin({
            postgresRequireSsl: yargs.requireSsl,
            potgresUri: yargs.postgresUri,
            username: yargs.username,
            password: yargs.password,
          });
        },
      )
      .command(
        'token <user>',
        'Gets a token for a user',
        (yargs) => {
          return yargs
            .positional('user', {
              demandOption: true,
              description: "The user's username or email address.",
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await getUserToken(yargs.postgresUri, yargs.requireSsl, yargs.user);
        },
      )
      .help();
  },

  async handler() {
    /* TODO */
  },
};
