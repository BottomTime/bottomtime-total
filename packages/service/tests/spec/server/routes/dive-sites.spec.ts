import { createMocks } from 'node-mocks-http';
import { faker } from '@faker-js/faker';
import { It, Mock, Times } from 'moq.ts';

import { createTestLogger } from '../../../test-logger';
import {
  DiveSite,
  DiveSiteManager,
  DiveSitesSortBy,
  SearchDiveSitesOptions,
} from '../../../../src/diveSites';
import {
  assertWritePermission,
  getDiveSite,
  loadDiveSite,
  searchDiveSites,
} from '../../../../src/server/routes/dive-sites';
import {
  ForbiddenError,
  MissingResourceError,
  UnauthorizedError,
  ValidationError,
} from '../../../../src/errors';
import { User } from '../../../../src/users';
import { SortOrder, UserRole } from '../../../../src/constants';

const Log = createTestLogger('dive-stie-routes');

describe('Dive Site Routes', () => {
  describe('Loading dive sites', () => {
    it('Will load a dive site from the ID in the request path', async () => {
      const siteId = faker.datatype.uuid();
      const diveSite = new Mock<DiveSite>()
        .setup((s) => s.id)
        .returns(siteId)
        .object();
      const diveSiteManager = new Mock<DiveSiteManager>()
        .setup((m) => m.getDiveSite(siteId))
        .returnsAsync(diveSite)
        .object();
      const { req, res } = createMocks({
        diveSiteManager,
        log: Log,
        params: { siteId },
      });
      const next = jest.fn();

      await loadDiveSite(req, res, next);

      expect(next).toBeCalledWith();
      expect(res._isEndCalled()).toBe(false);
      expect(req.selectedDiveSite).toBe(diveSite);
    });

    it('Will return a Not Found response if dive site ID cannot be found', async () => {
      const siteId = faker.datatype.uuid();
      const diveSiteManager = new Mock<DiveSiteManager>()
        .setup((m) => m.getDiveSite(siteId))
        .returnsAsync(undefined)
        .object();
      const { req, res } = createMocks({
        diveSiteManager,
        log: Log,
        params: { siteId },
      });
      const next = jest.fn();

      await loadDiveSite(req, res, next);

      expect(next.mock.lastCall[0]).toBeInstanceOf(MissingResourceError);
      expect(res._isEndCalled()).toBe(false);
    });

    it('Will return an error if an exception is thrown while retrieving dive site', async () => {
      const siteId = faker.datatype.uuid();
      const error = new Error('Fail!');
      const diveSiteManager = new Mock<DiveSiteManager>()
        .setup((m) => m.getDiveSite(siteId))
        .throwsAsync(error)
        .object();
      const { req, res } = createMocks({
        diveSiteManager,
        log: Log,
        params: { siteId },
      });
      const next = jest.fn();

      await loadDiveSite(req, res, next);

      expect(next).toBeCalledWith(error);
      expect(res._isEndCalled()).toBe(false);
    });
  });

  describe('Asserting write permission', () => {
    it('Will allow access if the user created the dive site', () => {
      const creator = faker.datatype.uuid();
      const user = new Mock<User>()
        .setup((u) => u.role)
        .returns(UserRole.User)
        .setup((u) => u.id)
        .returns(creator)
        .object();
      const selectedDiveSite = new Mock<DiveSite>()
        .setup((s) => s.creatorId)
        .returns(creator)
        .object();
      const { req, res } = createMocks({
        log: Log,
        selectedDiveSite,
        user,
      });
      const next = jest.fn();

      assertWritePermission(req, res, next);

      expect(next).toBeCalledWith();
      expect(res._isEndCalled()).toBe(false);
    });

    it('Will allow access if the user is an administrator', () => {
      const creator = faker.datatype.uuid();
      const user = new Mock<User>()
        .setup((u) => u.role)
        .returns(UserRole.Admin)
        .setup((u) => u.id)
        .returns(faker.datatype.uuid())
        .object();
      const selectedDiveSite = new Mock<DiveSite>()
        .setup((s) => s.creatorId)
        .returns(creator)
        .object();
      const { req, res } = createMocks({
        log: Log,
        selectedDiveSite,
        user,
      });
      const next = jest.fn();

      assertWritePermission(req, res, next);

      expect(next).toBeCalledWith();
      expect(res._isEndCalled()).toBe(false);
    });

    it('Will return a Forbidden error for non-admin users if they did not create the dive site', () => {
      const creator = faker.datatype.uuid();
      const user = new Mock<User>()
        .setup((u) => u.role)
        .returns(UserRole.User)
        .setup((u) => u.id)
        .returns(faker.datatype.uuid())
        .object();
      const selectedDiveSite = new Mock<DiveSite>()
        .setup((s) => s.creatorId)
        .returns(creator)
        .object();
      const { req, res } = createMocks({
        log: Log,
        selectedDiveSite,
        user,
      });
      const next = jest.fn();

      assertWritePermission(req, res, next);

      expect(next.mock.lastCall[0]).toBeInstanceOf(ForbiddenError);
      expect(res._isEndCalled()).toBe(false);
    });

    it('Will return an Unauthorized error for anonymous users', async () => {
      const creator = faker.datatype.uuid();
      const selectedDiveSite = new Mock<DiveSite>()
        .setup((s) => s.creatorId)
        .returns(creator)
        .object();
      const { req, res } = createMocks({
        log: Log,
        selectedDiveSite,
      });
      const next = jest.fn();

      assertWritePermission(req, res, next);

      expect(next.mock.lastCall[0]).toBeInstanceOf(UnauthorizedError);
      expect(res._isEndCalled()).toBe(false);
    });

    it('Will return a server error if no dive site is loaded', () => {
      const user = new Mock<User>()
        .setup((u) => u.role)
        .returns(UserRole.User)
        .setup((u) => u.id)
        .returns(faker.datatype.uuid())
        .object();
      const { req, res } = createMocks({
        log: Log,
        user,
      });
      const next = jest.fn();

      assertWritePermission(req, res, next);

      expect(next.mock.lastCall).toMatchSnapshot();
      expect(res._isEndCalled()).toBe(false);
    });
  });

  it('Will return a dive site as a JSON object', async () => {
    const expected = {
      id: 'abcd1234',
      name: 'Palancar Horseshoe',
      description: 'Epic reef',
    };
    const site = new Mock<DiveSite>()
      .setup((s) => s.toJSON())
      .returns(expected)
      .setup((s) => s.getCreator())
      .returnsAsync({
        id: 'user1',
        username: 'YannTheMann',
        displayName: 'Yann Jenkins',
        memberSince: new Date('2023-08-22T20:41:27.869Z'),
      });
    const { req, res } = createMocks({
      log: Log,
      selectedDiveSite: site.object(),
    });

    await getDiveSite(req, res);

    expect(res._isEndCalled()).toBe(true);
    expect(res._getJSONData()).toEqual(expected);
    expect(res._getStatusCode()).toBe(200);
    expect(site.verify((s) => s.getCreator(), Times.Once()));
  });

  describe('Searching Dive Sites', () => {
    let searchResults: DiveSite[];

    beforeAll(() => {
      searchResults = new Array<DiveSite>(10);
      for (let i = 0; i < searchResults.length; i++) {
        searchResults[i] = new Mock<DiveSite>()
          .setup((s) => s.id)
          .returns(`expected-${i}`)
          .setup((s) => s.toSummaryJSON())
          .returns({
            id: `expected-${i}`,
          })
          .object();
      }
    });

    it('Will perform a basic query with no options', async () => {
      const diveSiteManager = new Mock<DiveSiteManager>()
        .setup((m) =>
          m.searchDiveSites(It.IsAny<SearchDiveSitesOptions | undefined>()),
        )
        .returnsAsync(searchResults);
      const { req, res } = createMocks({
        diveSiteManager: diveSiteManager.object(),
        log: Log,
      });
      const next = jest.fn();

      await searchDiveSites(req, res, next);

      expect(next).not.toBeCalled();
      expect(res._isEndCalled()).toBe(true);
      expect(res._getStatusCode()).toBe(200);
      expect(res._getJSONData()).toEqual({
        results: searchResults.length,
        sites: searchResults.map((result) => result.toSummaryJSON()),
      });
      diveSiteManager.verify(
        (m) =>
          m.searchDiveSites(
            It.Is<SearchDiveSitesOptions>(
              (opts) => JSON.stringify(opts) === '{}',
            ),
          ),
        Times.Once(),
      );
    });

    it('Will perform a query with many options', async () => {
      const diveSiteManager = new Mock<DiveSiteManager>()
        .setup((m) =>
          m.searchDiveSites(It.IsAny<SearchDiveSitesOptions | undefined>()),
        )
        .returnsAsync(searchResults);
      const expected = JSON.stringify({
        query: 'lake',
        location: {
          lat: 20.3460942599771,
          lon: -87.0260668489514,
        },
        radius: 80,
        freeToDive: false,
        shoreAccess: true,
        sortBy: DiveSitesSortBy.Rating,
        sortOrder: SortOrder.Descending,
        skip: 400,
        limit: 100,
      });
      const { req, res } = createMocks({
        diveSiteManager: diveSiteManager.object(),
        log: Log,
        query: {
          query: 'lake',
          location: '20.3460942599771,-87.0260668489514',
          radius: '80',
          freeToDive: 'false',
          shoreAccess: 'true',
          sortBy: 'rating',
          sortOrder: 'desc',
          skip: '400',
          limit: '100',
        },
      });
      const next = jest.fn();

      await searchDiveSites(req, res, next);

      expect(next).not.toBeCalled();
      expect(res._isEndCalled()).toBe(true);
      expect(res._getStatusCode()).toBe(200);
      expect(res._getJSONData()).toEqual({
        results: searchResults.length,
        sites: searchResults.map((result) => result.toSummaryJSON()),
      });
      diveSiteManager.verify((m) =>
        m.searchDiveSites(It.Is((opts) => JSON.stringify(opts) === expected)),
      );
    });

    it('Will return a ValidationError if query string is invalid', async () => {
      const diveSiteManager = new Mock<DiveSiteManager>().object();
      const { req, res } = createMocks({
        diveSiteManager: diveSiteManager,
        log: Log,
        query: {
          location: '-300.123,504.31432',
        },
      });
      const next = jest.fn();

      await searchDiveSites(req, res, next);

      expect(next.mock.lastCall[0]).toBeInstanceOf(ValidationError);
      expect(res._isEndCalled()).toBe(false);
    });

    it('Will pass along an error if the DiveSiteManager throws an exception', async () => {
      const error = new Error('Uh oh!');
      const diveSiteManager = new Mock<DiveSiteManager>()
        .setup((m) =>
          m.searchDiveSites(It.IsAny<SearchDiveSitesOptions | undefined>()),
        )
        .throwsAsync(error);
      const { req, res } = createMocks({
        diveSiteManager: diveSiteManager.object(),
        log: Log,
      });
      const next = jest.fn();

      await searchDiveSites(req, res, next);

      expect(next).toBeCalledWith(error);
      expect(res._isEndCalled()).toBe(false);
    });
  });
});
