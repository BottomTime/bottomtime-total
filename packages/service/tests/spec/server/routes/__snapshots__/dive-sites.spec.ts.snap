// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dive Site Routes Asserting write permission Will return a server error if no dive site is loaded 1`] = `
[
  [Error: No dive site loaded!],
]
`;
